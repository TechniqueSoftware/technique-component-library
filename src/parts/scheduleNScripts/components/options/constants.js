export const SCRIPT_CONTACT_TYPES = {
  EMAIL: 2,
  SMS: 3,
  NONE: 4
}
