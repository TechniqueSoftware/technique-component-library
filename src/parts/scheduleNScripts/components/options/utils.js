export const followUpTimeToObject = followUpTime => {
  const getHour = h => {
    if (h === '00') return 12
    const parsedHour = parseInt(h, 10)
    return parsedHour > 12 ? parsedHour - 12 : parsedHour
  }
  const time = followUpTime.split(':')
  const [oldHour, oldMinute] = time
  const hour = getHour(oldHour)
  const minute = parseInt(oldMinute, 10)
  const meridiem = oldHour >= 12 ? 'PM' : 'AM'
  return { hour, minute, meridiem }
}

export const objectToFollowUpTime = (hour, minute, meridiem) => {
  const parseHour = h => meridiem === 'PM' ? parseInt(h, 10) + 12 : h
  return `${parseHour(hour === 12 ? '00' : hour)}:${minute === 0 ? '00' : minute}:00`
}
