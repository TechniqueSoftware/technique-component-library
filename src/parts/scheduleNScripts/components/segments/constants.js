export const FOLLOWUP_STATUS = {
  APPT_BOOKED: 2,
  MEMBER: 6,
  WEB_LEAD: 11
}
