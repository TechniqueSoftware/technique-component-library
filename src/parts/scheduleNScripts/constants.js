export const TP_ACTION_TYPES = {
  CALL: 1,
  EMAIL: 2,
  TEXT: 3
}

export const SCRIPT_TABNAME = [{
  id: '1',
  name: 'Call',
  className: 'BsTelephone'
}, {
  id: '2',
  name: 'Email',
  className: 'BsEnvelope'
}, {
  id: '3',
  name: 'Text',
  className: 'BsChatDots'
}]

export const SEGMENT_TABNAME = [
  {
    id: '0',
    name: 'Manage Segments'
  },
  {
    id: '1',
    name: 'Copy Segments'
  }
]

export const SNACKBAR_MESSAGES = {
  NEW_SUCCESS: 'Touchpoint has been created successfully!',
  EDIT_SUCCESS: 'Touchpoint has been updated!',
  DELETE_SUCCESS: 'Touchpoint has been deleted successfully!',
  SCRIPT_SUCCESS: 'Script has been saved successfully!',
  SEGMENT_NEW: 'Segment has been created successfully!',
  SEGMENT_EDIT: 'Changes for this segment have been saved successfully!',
  SEGMENT_DELETE: 'Segment has been deleted successfully!'
}

export const STATUSES_WITH_SEGMENTS = [
  { type: 1, id: 11, attributeType: 'Marketing Source', attributeId: 'marketing' }, // Prospecting -> Web Lead
  { type: 1, id: 2, attributeType: 'Marketing Source', attributeId: 'marketing' }, // Prospecting -> Appt Booked
  { type: 1, id: 6, attributeType: 'Membership Type', attributeId: 'membership' } // Member -> Member
]

export const PERMISSION_KEYS = {
  SCHEDULES: {
    1: 'canEditMemberSalesSchedules',
    2: 'canEditPtSchedules',
    3: 'canEditStudioSchedules'
  },
  SCRIPTS: {
    1: 'canEditMemberSalesScripts',
    2: 'canEditPtScripts',
    3: 'canEditStudioScripts'
  }
}
