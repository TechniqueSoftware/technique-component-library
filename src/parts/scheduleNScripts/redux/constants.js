export const TP_OPTIONS = {
  MENU: 'menu',
  NEW: 'new',
  EDIT: 'edit'
}

export const GROUP_NAMES = {
  PROSPECT: 'Prospect',
  ACTIVE_MEMBER: 'Active Member',
  INACTIVE_MEMBER: 'Inactive Member',
  PT_PROSPECT: 'PT Prospect',
  PT_MEMBER: 'PT Member',
  PT_OPT_OUT: 'PT Opt Out',
  PERSONAL_TRAINING: 'Personal Training'
}
