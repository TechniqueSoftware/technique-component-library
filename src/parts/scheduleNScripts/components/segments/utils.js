import { FOLLOWUP_STATUS } from './constants'

export const parseTitle = followUpStatus => {
  switch (followUpStatus) {
    case FOLLOWUP_STATUS.WEB_LEAD:
      return 'Web Lead'
    case FOLLOWUP_STATUS.APPT_BOOKED:
      return 'Appt Booked'
    case FOLLOWUP_STATUS.MEMBER:
      return 'Member'
  }
}

export const parseLocations = segmentLocations => segmentLocations.map(({ locationName, locationId }) => ({
  label: `${locationName}`,
  value: locationId
}))

export const checkPermission = userPermission => userPermission.some(
  element => element.permissionName === 'CopySegmentsAction' && element.permissionEffect === 'ALLOW')
