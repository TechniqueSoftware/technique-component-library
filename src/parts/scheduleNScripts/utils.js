import { PERMISSION_KEYS, STATUSES_WITH_SEGMENTS } from './constants'

export const checkIfSegmentStatus = (type, id) => {
  const hasSegmentation = !!STATUSES_WITH_SEGMENTS.find(s => s.type === parseInt(type, 10) && s.id === parseInt(id, 10))
  return hasSegmentation
}

const setSchedulesDayNum = schedules => schedules.reduce(
  (acc, curr) => {
    const dayNum = curr.daysTillNext + acc.dayNum
    return { schedules: acc.schedules.concat({ ...curr, dayNum }), dayNum }
  },
  { dayNum: 0, schedules: [] }
).schedules

export const getStatusSegments = (type, id, segments) => {
  if (!checkIfSegmentStatus(type, id)) return []
  const currentStatus = segments.find(
    ({ followUpType, followUpStatus }) => followUpType === parseInt(type, 10) && followUpStatus === parseInt(id, 10)
  )
  if (!currentStatus) return []
  const { followUpSegments = [], followUpSchedules = [], followUpSegmentScripts = [] } = currentStatus
  return followUpSegments.map(segment => {
    const schedulesWithScripts = followUpSchedules.reduce((acc, schedule) => {
      if (segment.segmentId !== schedule.segmentId) return acc
      return acc.concat({
        ...schedule,
        scripts: followUpSegmentScripts.filter(
          script => script.followUpSeqNum === schedule.followUpSeqNum && script.segmentId === schedule.segmentId
        )
      })
    }, [])
    return { ...segment, followUpSchedules: setSchedulesDayNum(schedulesWithScripts) }
  })
}

export const sortSegments = segments => [...segments].sort((a, b) => a.name.localeCompare(b.name))

// Formatting club-os data into something usable
export const parseSnSData = (data, followUpSchedules, followUpScripts, segmentsData, isClub) => {
  const schedulesPermissions = followUpSchedules.permissions
  const scriptsPermissions = followUpScripts.permissions
  const schedules = isClub ? followUpSchedules : followUpSchedules.schedules
  const scripts = isClub ? followUpScripts : followUpScripts.scripts

  const group = (acc, curr) => {
    const { followUpStatus, followUpType } = curr
    const key = `${followUpStatus}-${followUpType}`
    if (acc[key]) acc[key].push(curr)
    else acc[key] = [curr]
    return acc
  }

  const sch = schedules.reduce(group, {})
  const scr = scripts.reduce(group, {})

  Object.keys(sch).forEach(key => { sch[key] = setSchedulesDayNum(sch[key]) })

  Object.keys(scr).forEach(key => {
    const scripts = scr[key]
    scripts.forEach(script => {
      const schedule = sch[key]?.find(el => el.followUpSeqNum === script.followUpSeqNum)
      if (!schedule) return
      if (schedule.scripts) schedule.scripts.push(script)
      else schedule.scripts = [script]
    })
  })

  const followups = data.map(el => {
    const { type, statuses, id } = el
    const canEditSchedules = isClub ? undefined : schedulesPermissions[PERMISSION_KEYS.SCHEDULES[type]]
    const canEditScripts = isClub ? undefined : scriptsPermissions[PERMISSION_KEYS.SCRIPTS[type]]
    const newStatuses = statuses.map(st => {
      const key = `${st.id}-${type}`
      const unsortedSegments = getStatusSegments(type, st.id, segmentsData)
      const segments = sortSegments(unsortedSegments)
      return { ...st, followUpTab: id, schedules: sch[key] || [], segments }
    })
    return { ...el, statuses: newStatuses, canEditSchedules, canEditScripts }
  })

  return {
    followups,
    canEditSchedules: followups[0].canEditSchedules,
    canEditScripts: followups[0].canEditScripts
  }
}

export const getSchedulesByDay = (schedules, dayNum, order) => {
  const { earlierSch, sameDaySch, laterSch } = schedules.reduce(
    (acc, sch) => {
      if (sch.dayNum < dayNum) acc.earlierSch = acc.earlierSch.concat(sch)
      if (sch.dayNum === dayNum) acc.sameDaySch = acc.sameDaySch.concat(sch)
      if (sch.dayNum > dayNum) acc.laterSch = acc.laterSch.concat(sch)
      return acc
    },
    { earlierSch: [], sameDaySch: [], laterSch: [] }
  )
  return {
    earlierSch: order === 'first' ? earlierSch : earlierSch.concat(sameDaySch),
    laterSch: order === 'first' ? sameDaySch.concat(laterSch) : laterSch
  }
}

export const getCurrentSchedules = (segmentId, status) => {
  if (!segmentId) return status.schedules
  return status.segments.find(segment => segment.segmentId === segmentId).followUpSchedules
}

export const getStatus = (followups, type, status) => {
  const fs = followups.filter(f => f.type === type)
  if (!fs.length) return null

  let st = null
  for (let i = 0; i < fs.length; i++) {
    st = fs[i].statuses.find(s => s.id === `${status}`)
    if (st) break
  }
  return st
}

export const setFollowUpStatusValue = (followUps, status, value) => {
  const newFollowups = followUps.map(followUp => {
    if (followUp.id !== status.followUpTab) return followUp
    const newStatuses = followUp.statuses.map(st => {
      if (st.id !== status.id) return st
      return { ...st, ...value }
    })
    return { ...followUp, statuses: newStatuses }
  })
  return newFollowups
}

export const setFollowUpScripts = (followUps, { followUpType, followUpStatus }, scriptsData, segmentId) => {
  const status = getStatus(followUps, followUpType, followUpStatus)
  const currentSchedules = getCurrentSchedules(segmentId, status)
  const schedules = currentSchedules.map(schedule => {
    if (schedule.followUpSeqNum !== scriptsData.followUpSeqNum) return schedule
    return scriptsData
  })
  const statusValueToUpdate = !segmentId
    ? { schedules }
    : {
      segments: status.segments.map(
        segment => (segment.segmentId !== segmentId ? segment : { ...segment, followUpSchedules: schedules }) // eslint-disable-line max-len
      )
    }
  return setFollowUpStatusValue(followUps, status, statusValueToUpdate)
}

export const updateTouchpointInFollowups = (followUps, tp, data) => {
  const { followUpType, followUpStatus, followUpSeqNum } = tp
  const status = getStatus(followUps, followUpType, followUpStatus)
  const newSchedules = status.schedules.map(sch => {
    if (sch.followUpSeqNum !== followUpSeqNum) return sch
    const newSch = { ...sch, ...data }
    return newSch
  })
  return setFollowUpStatusValue(followUps, status, { schedules: newSchedules })
}
