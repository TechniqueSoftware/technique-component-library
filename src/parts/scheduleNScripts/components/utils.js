import { STATUSES_WITH_SEGMENTS } from '@parts/scheduleNScripts/constants'
// SEGMENTS
export const checkIfAttributeIsInUse = (segments, selectedAttributes, valueToCheck, segmentId) => {
  if (valueToCheck === '') return false
  const listToCheck = segments.reduce(
    (acc, segment) => {
      if (segmentId && segment.segmentId === segmentId) return acc
      return acc.concat(segment.segmentAttributes)
    },
    selectedAttributes.map(({ value }) => ({ attributeValue: value }))
  )
  return !!listToCheck.find(({ attributeValue }) => attributeValue === valueToCheck)
}

export const getSegmentStatusProp = ({ followUpType, followUpStatus }, keyword) => {
  const segmentTp = STATUSES_WITH_SEGMENTS.find(
    s => s.type === parseInt(followUpType, 10) && s.id === parseInt(followUpStatus, 10)
  )
  return segmentTp?.[keyword]
}

export const parseAttributes = (attributes, type) => attributes.reduce((acc, attr) => {
  if (!attr.value) return acc
  return acc.concat({ attributeValue: attr.value, attributeType: type, attributeId: attr.id })
}, [])

export const unparseAttributes = attributes => attributes.map(({ attributeValue, attributeId }) => ({
  name: attributeValue,
  value: attributeValue,
  id: attributeId
}))
