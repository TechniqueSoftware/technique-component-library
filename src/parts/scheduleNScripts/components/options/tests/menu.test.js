import React from 'react'
import * as redux from 'react-redux'
import { render, screen } from '@testing-library/react'
import { TestProvider as Provider } from '@redux/utils'

import { TP_OPTIONS } from '../../../redux'
import { TP_ACTION_TYPES } from '../../../constants'
import Menu from '..'

const mockDispatch = jest.fn()
jest.spyOn(redux, 'useDispatch').mockReturnValue(mockDispatch)

describe('<New />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
      scheduleNScripts: {
        touchpointCrud: {
          action: TP_OPTIONS.MENU,
          data: { displayName: 'Some status here' }
        }
      }
    }))
    const { asFragment } = render(
      <Provider>
        <Menu />
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  describe('New Touchpoint', () => {
    beforeAll(() => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
        scheduleNScripts: {
          touchpointCrud: {
            action: TP_OPTIONS.NEW,
            data: { displayName: 'Some status here' }
          }
        }
      }))
    })

    it('renders title', () => {
      render(
        <Provider>
          <Menu />
        </Provider>
      )

      expect(screen.getByText('Touchpoint:')).toBeInTheDocument()
      expect(screen.getByText('Some status here')).toBeInTheDocument()
    })

    it('renders content', () => {
      render(
        <Provider>
          <Menu />
        </Provider>
      )

      expect(screen.getByText('Immediate')).toBeInTheDocument()
    })
  })

  describe('Edit Schedule', () => {
    beforeAll(() => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
        scheduleNScripts: {
          touchpointCrud: {
            action: TP_OPTIONS.EDIT,
            data: {
              followUpSeqNum: 1,
              daysTillNext: 3,
              displayName: 'Day 4'
            }
          }
        }
      }))
    })

    it('renders content', () => {
      render(
        <Provider>
          <Menu />
        </Provider>
      )

      expect(screen.queryByText('Touchpoint:')).toBeInTheDocument()
      expect(screen.getByText('Day 4')).toBeInTheDocument()
      expect(screen.queryByText('Save')).toBeInTheDocument()
    })
  })

  describe('Edit Touchpoint', () => {
    beforeAll(() => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
        scheduleNScripts: {
          touchpointCrud: {
            action: TP_OPTIONS.MENU,
            data: {
              followUpSeqNum: 1,
              daysTillNext: 3,
              rowName: 'Web Lead',
              displayName: 'Day 4'
            }
          }
        }
      }))
    })

    it('renders title', () => {
      render(
        <Provider>
          <Menu />
        </Provider>
      )

      expect(screen.queryByText('Edit:')).toBeInTheDocument()
      expect(screen.getByText('Web Lead Day 4')).toBeInTheDocument()
    })
  })

  describe('Render model windows', () => {
    beforeAll(() => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
        scheduleNScripts: {
          touchpointCrud: {
            action: TP_ACTION_TYPES.CALL,
            data: {
              followUpSeqNum: 1,
              daysTillNext: 3,
              rowName: 'Web Lead',
              displayName: 'Day 4',
              scripts: []
            }
          }
        }
      }))
    })

    it('renders Configure Follow Up', () => {
      render(
        <Provider>
          <Menu />
        </Provider>
      )
      expect(screen.queryByText('Configure Follow Up Contact')).toBeInTheDocument()
    })
  })
})
