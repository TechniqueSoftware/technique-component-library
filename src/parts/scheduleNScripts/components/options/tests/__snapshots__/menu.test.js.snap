// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<New /> should match snapshot 1`] = `<DocumentFragment />`;
